lstchain/image/muon/muon_analysis.py

```diff
@@ -131,90 +131,132 @@ def analyze_muon_event(event_id, image, geom, equivalent_focal_length,
+    pix_ring_2 = image[dist_mask_2]
+
+#    nom_dist = np.sqrt(np.power(muonringparam.ring_center_x,2) 
+#                    + np.power(muonringparam.ring_center_y, 2))
 
     muonringparam.ring_containment = ring_containment(
             muonringparam.ring_radius,
             cam_rad,
             muonringparam.ring_center_x,
             muonringparam.ring_center_y)
 
-    ctel = MuonLineIntegrate(
-                mirror_radius, hole_radius = 0.308 * u.m,
-                pixel_width=0.1 * u.deg,
-                sct_flag=False,
-                secondary_radius = 0. * u.m
-            )
+    
+    radial_distribution = radial_light_distribution(
+        muonringparam.ring_center_x,
+        muonringparam.ring_center_y,
+        x[clean_mask], y[clean_mask],
+        image[clean_mask])
+    
+    
+    # Do complicated calculations (minuit-based max likelihood ring fit) only for selected rings:
+    candidate_clean_ring = all(
+        [radial_distribution['standard_dev'] < max_radial_stdev,
+         radial_distribution['excess_kurtosis'] < max_radial_excess_kurtosis,
+         npix_above_threshold(pix_ring, tailcuts[0]) > min_pix_fraction_after_cleaning * min_pix,
+         npix_composing_ring(pix_ring) > min_pix,
+         muonringparam.ring_radius < max_ring_radius,
+         muonringparam.ring_radius > min_ring_radius
+        ])
+
+    if  candidate_clean_ring:
+        ctel = MuonLineIntegrate(
+            mirror_radius, hole_radius = 0.308 * u.m,
+            pixel_width=0.1 * u.deg,
+            sct_flag=False,
+            secondary_radius = 0. * u.m
+        )
+
+        muonintensityoutput = ctel.fit_muon(
+            muonringparam.ring_center_x,
+            muonringparam.ring_center_y,
+            muonringparam.ring_radius,
+            x[dist_mask], y[dist_mask],
+            image[dist_mask])
+
+        dist_ringwidth_mask = np.abs(dist - muonringparam.ring_radius) < (muonintensityoutput.ring_width)
+        # We do the calculation of the ring completeness (i.e. fraction of whole circle) using the pixels
+        # within the "width" fitted using MuonLineIntegrate.
+        muonintensityoutput.ring_completeness = ring_completeness(
+            x[dist_ringwidth_mask], y[dist_ringwidth_mask], image[dist_ringwidth_mask],
+            muonringparam.ring_radius,
+            muonringparam.ring_center_x,
+            muonringparam.ring_center_y,
+            threshold=30,
+            bins=30)
+
+        pix_ringwidth_im = image * dist_ringwidth_mask
+        idx_ringwidth = np.nonzero(pix_ringwidth_im)
+        muonintensityoutput.ring_pix_completeness = npix_above_threshold(pix_ringwidth_im[idx_ringwidth], tailcuts[0]) / len(
+            pix_ring[idx_ringwidth])
+
+    else:
+            muonintensityoutput = MuonIntensityParameter()
+            # Set default values for cases in which the muon intensity fit (with MuonLineIntegrate) is not done:
```

Comment:
Passing the container fields as keyword arguments to the constructor is faster, as the container won't initialize defaults and then override them.

However, I will also open a PR in ctapipe, that the default values are actually `np.nan`